src/pgwire/src/protocol.rs

```diff
@@ -180,22 +174,42 @@ where
             }
         };
         let admin_role = frontegg.admin_role();
-        let external_metadata_rx = session.retain_external_metadata_transmitter();
         match frontegg
             .exchange_password_for_token(&password)
             .await
             .and_then(|token| {
-                frontegg.continuously_validate_access_token(token, user.clone(), move |claims| {
-                    let external_metadata = convert_claims_to_external_metadata(claims, admin_role);
-                    // Ignore error if client has hung up.
-                    let _ = external_metadata_rx.send(external_metadata);
-                })
+                // Create a session based on the validated claims.
+                //
+                // In particular, it's important that the username come from the
+                // claims, as Frontegg may return an email address with
+                // different casing than the user supplied via the pgwire
+                // username field. We want to use the Frontegg casing as
+                // canonical.
+                let claims = frontegg.validate_access_token(&token.access_token, Some(&user))?;
```

Comment:
This line is the first line in continuously_validate_access_token, which immediately calls claims_processor with its result. I think it might make more sense change continuously_validate_access_token to do exactly what we need here, or else have the claims_processor Fn do something like send its first claim on a channel.

Reply:
Yes, it does some duplicate work. I tried a few refactorings along those lines, but my attempts ended up with more `expect`s. 

What do you think about having `continuously_validate_access_token` _not_ do any initial validation of the claim; instead it would only start the background task and validate future refreshes.

Reply:
That sounds ok? Does it increase safety if it also takes as an argument the first expected claim? I'm worried about some race condition where something changes between first and second checks, and we didn't verify something? I'm making this up.

Reply:
Refactoring this was shockingly difficult, but I finally have something I'm really happy with. PTAL. The revalidation function now _only_ does revalidation, but I refactored the API with a bunch of types that statically ensure that the caller can't forget to do the initial validation.